### DISCUSSION TITLE
[SUGGESTION]  unsafe_narrow is not a good name for forced upcasts.

### DISCUSSION THREAD

**Discussion by MaxSagebaum** [hsutter/cppfront#630] [Suggestions]
If you have e.g. expression templates, then the upcast to the implementing class needs to be done with `unsafe_narrow`.

Maybe an alias of `unsafe_narrow` like `unsafe_cast`, 'unsafe_pointer_cast` or something like that could be introduced.

Example: https://cpp2.godbolt.org/z/jb1MY3Mcq
```
Exp : <Impl> type = {  
  cast: (this) -> forward Impl = cpp2::unsafe_narrow<Impl>(this&)*; // unsafe_narrow does not really express what this cast is.
}

A : type = {
  this: Exp<A> = ();
  foo: (this) = std::cout << "foo" << std::endl;
}

main: () -> int = {
  a: A = ();
  e: *Exp<A> = a&;

  e*.cast().foo();

  return 0;
}
```

**Comment by JohelEGP** [comment]
I call this `static_downcast`: <https://johelegp.github.io/jegp/static.downcast>.
Inspired by <https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-named>,
I created my own more specific one.

**Comment by MaxSagebaum** [comment]
Yes, that is also possible.

**Comment by JohelEGP** [comment]
By the way,
I noticed in your other issue (and this one)
that you mixed up the terms upcast and downcast.
Then, I tried to find it in the specification for `static_cast`,
but it wasn't mentioned, so I just left it unsaid.
Now, I did find other examples.
These use "upcast" when converting from derived to base:
<https://eel.is/c++draft/class.cdtor#example-3> and <https://eel.is/c++draft/class.cdtor#example-1>.

**Comment by MaxSagebaum** [comment]
Thanks for pointing this out. I knew, that this can also be seen the other way round. I will start using it in terms of the type system and not the functionality.

**Comment by JohelEGP** [comment]
`cpp2::unsafe_narrow` is modeled after [`gsl::narrow_cast`][].
[`gsl::narrow_cast`][] _seems_ to be modeled after [`gsl::narrow`][]
based on its uses in the [C++ Core Guidelines][].
Although `gsl::narrow_cast<T>(x)` is specified as being `static_cast<T>(x)`
[ _Note_: `static_cast<T>(x)` is illegal Cpp2. -- _end note_ ]
it's not clear if those should work with non-arithmetic types:
See <https://github.com/isocpp/CppCoreGuidelines/issues/1845> and linked issues (and their linked issues/PRs).

[`gsl::narrow_cast`]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#SS-utilities
[`gsl::narrow`]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#SS-utilities
[C++ Core Guidelines]: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines

**Comment by MaxSagebaum** [comment]
Ok, `cpp2::unsafe_narrow` should be equally handled as the gsl one. But there should be additional casts in cpp2.
